src/test/java/org/prebid/server/analytics/reporter/greenbids/GreenbidsAnalyticsReporterTest.java

```diff
@@ -0,0 +1,643 @@
+                jsonCaptor.capture(),
+                eq(greenbidsAnalyticsProperties.getTimeoutMs()));
+
+        final String capturedJson = jsonCaptor.getValue();
+        final CommonMessage capturedCommonMessage = jacksonMapper.mapper()
+                .readValue(capturedJson, CommonMessage.class);
+        assertThat(capturedCommonMessage).usingRecursiveComparison()
+                .ignoringFields("billingId", "greenbidsId")
+                .isEqualTo(expectedCommonMessage);
+        assertThat(capturedCommonMessage.getGreenbidsId()).isNotNull();
+        assertThat(capturedCommonMessage.getBillingId()).isNotNull();
+
+        assertThat(headersCaptor.getValue().get(HttpUtil.ACCEPT_HEADER))
+                .isEqualTo(HttpHeaderValues.APPLICATION_JSON.toString());
+        assertThat(headersCaptor.getValue().get(HttpUtil.CONTENT_TYPE_HEADER))
+                .isEqualTo(HttpHeaderValues.APPLICATION_JSON.toString());
+    }
+
+    @Test
+    public void shouldFailWhenBidResponseIsNull() {
+        // given
+        final AuctionContext auctionContext = givenAuctionContextWithNoBidResponse();
+        final AuctionEvent event = AuctionEvent.builder()
+                .auctionContext(auctionContext)
+                .bidResponse(auctionContext.getBidResponse())
+                .build();
+
+        // when
+        final Future<Void> result = target.processEvent(event);
+
+        // then
+        assertThat(result.failed()).isTrue();
+        assertThat(result.cause())
+                .hasMessageStartingWith("Bid response or auction context cannot be null");
+    }
+
+    @Test
+    public void shouldFailWhenAuctionContextIsNull() {
+        // given
+        final AuctionEvent event = mock(AuctionEvent.class);
+        when(event.getAuctionContext()).thenReturn(null);
+
+        // when
+        final Future<Void> result = target.processEvent(event);
+
+        // then
+        assertThat(result.failed()).isTrue();
+        assertThat(result.cause())
+                .hasMessageStartingWith("Bid response or auction context cannot be null");
+    }
+
+    @Test
+    public void shouldFailOnEncodeException() {
+        // given
+        final Banner banner = givenBanner();
+        final Imp imp = Imp.builder()
+                .banner(banner)
+                .build();
+        final AuctionContext auctionContext = givenAuctionContext(imp);
+        final AuctionEvent event = AuctionEvent.builder()
+                .auctionContext(auctionContext)
+                .bidResponse(auctionContext.getBidResponse())
+                .build();
+
+        final GreenbidsJacksonMapper mockJacksonMapper = mock(GreenbidsJacksonMapper.class);
+        final ObjectMapper realObjectMapper = new ObjectMapper();
+        when(mockJacksonMapper.mapper()).thenReturn(realObjectMapper);
+        doThrow(new EncodeException("Failed to encode as JSON")).when(mockJacksonMapper).encodeToString(any());
+
+        target = new GreenbidsAnalyticsReporter(
```

Comment:
Global test reference should not be changed inside tests, pls fix it

Reply:
The global target, properties and jackson mapper are defined globally in the `setUp`

```java
@Before
    public void setUp() {
        final ObjectMapper mapper = new ObjectMapper();
        jacksonMapper = new GreenbidsJacksonMapper(mapper);

        greenbidsAnalyticsProperties = GreenbidsAnalyticsProperties.builder()
                .exploratorySamplingSplit(0.9)
                .analyticsServerVersion("2.2.0")
                .analyticsServerUrl("http://localhost:8080")
                .configurationRefreshDelayMs(10000L)
                .timeoutMs(100000L)
                .build();

        target = new GreenbidsAnalyticsReporter(
                greenbidsAnalyticsProperties,
                jacksonMapper,
                httpClient,
                clock,
                prebidVersionProvider);
    }
```

Concerning this specific test `shouldFailOnEncodeException` we had the discussion with Anton that to simulate the case when json encoding is failed we need to mock jackson mapper instead of using the global mapper and then redefine the target `GreenbidsAnalyticsReporter` for it with this new mapper. This is the only case where I had to redefine the target inside the test

Reply:
ok, agree